### ISSUE TITLE
Telegram Desktop 5.3.2 on Rocky Linux 9 does not display Chinese characters in messages

### ISSUE THREAD

**Issue by hg2581** [telegramdesktop/tdesktop#28229]
### Steps to reproduce

Rocky Linux 9, English as OS language and fcitx5 to input both simplified and traditional characters in applications. I can user fcitx5 to type Chinese characters and they show up correct in the input module but are rendered as squares in Telegram. This is true also for received messages containing Chinese characters.

How do I fix this? I do not want to change the interface language from English.

### Expected behaviour

Chinese characters should be rendered correctly.

### Actual behaviour

Chinese characters are rendered as squares.

### Operating system

Rocky Linux 9

### Version of Telegram Desktop

5.3.2

### Installation source

Static binary from official website

**Comment by ilya-fedin** [comment]
> How do I fix this?

Can you reproduce with other Qt applications like qBittorrent?

**Comment by hg2581** [comment]
I am not running qBittorrent. Do you have any other suggestions for Qt-based applications I could try?

Can you reproduce it on one of your setups?

**Comment by Aokromes** [comment]
tried to view a random message https://t.me/chinan/9 on fedora 40 and works for me.

**Comment by ilya-fedin** [comment]
> I am not running qBittorrent.

You can install it just for a test

> Do you have any other suggestions for Qt-based applications I could try?

No, I can't remember any.

> Can you reproduce it on one of your setups?

Well, I don't have Rocky Linux setups but the ones I have: it's either no application on the system could show chinese or all of them including tdesktop. I don't have chinese input methods though so idk whether it could show chinese even when applications can't...

**Comment by hg2581** [comment]
> tried to view a random message https://t.me/chinan/9 on fedora 40 and antiX-23.1 and works for me.

This opens fine in my browser but not in Telegram Desktop where every Chinese character is rendered as a square. It also opens fine in my Android Telegram although that is unrelated of course.

Rocky LInux 9 uses SELinux but I do not see any AVC violations so that does not seem to be the problem.

**Comment by ilya-fedin** [comment]
We need to know whether it happens in qBittorent on your PC

**Comment by hg2581** [comment]
> We need to know whether it happens in qBittorrent on your PC

I just downloaded qbittorrent 4.6.3 from epel and tried to enter Chinese characters in the "Filter torrents" input box but although I can use fcitx5 to switch between US, a European language and Chinese, I am unable to even enter Chinese characters here. Non-ASCII works fine.

Is qbittorrent using the same version of Qt as Telegram? Also, are you packaging Telegram in some special way that might interfere with accessing fonts on the host computer?

**Comment by ilya-fedin** [comment]
> Is qbittorrent using the same version of Qt as Telegram?

I'm not sure as you say you have installed qBittorent from epel. tdesktop static binary uses Qt 6.7.2. IIRC qBittorent shows that information in about dialog. qBittorrent has AppImages on their official website, it seem to use Qt 6.6.3.

**Comment by ilya-fedin** [comment]
> I am unable to even enter Chinese characters here.

couldn't you copy-paste them?

**Comment by hg2581** [comment]
> > I am unable to even enter Chinese characters here.
> 
> couldn't you copy-paste them?

Good question. I just tried by copying some text from the link Aokromes sent and I can paste them into both qbittorrent and into Telegram Desktop but in both instances they are rendered as squares. Copying/pasting does /not/ involve fcitx5 of course.

I did forget to mention that yesterday I wanted to see if I could switch the Telegram Desktop interface language to Chinese but that is not one of the options in settings. Korean is however, and that language choice in the settings menu is rendered as squares just under the English-language menu item name.

The problem therefore seems to be with Telegram and has nothing to do with fcitx5 or inputting Chinese characters in themselves.

My uneducated hypothesis is that Telegram does not properly pick up CJK fonts installed on the host OS?

**Comment by ilya-fedin** [comment]
Well, Telegram doesn't manage fonts, it's done by the underlying toolkit (Qt). And, as you can see, Qt is unable to see Chinese fonts on your system. There's sadly nothing tdesktop could do about that.

**Comment by hg2581** [comment]
> Well, Telegram doesn't manage fonts, it's done by the underlying toolkit (Qt). And, as you can see, Qt is unable to see Chinese fonts on your system. There's sadly nothing tdesktop could do about that.

Well, the bug is in Telegram regardless of which underlying toolkit it uses. I am an end-user of Telegram and not a developer. Since your company is using Qt for your application, you should raise the issue with them. Perhaps Qt 6.7 relies on the presence of some other software on my system that it assumed were installed but is not? This is a fairly fresh installation about a month old of RL9.

Chinese character input and rendering work flawlessly in the other applications I need to do it in, .e.g Chrome, LibreOffice, geany editor, evolution, terminal windows to name the ones I use.

**Comment by ilya-fedin** [comment]
> Well, the bug is in Telegram regardless of which underlying toolkit it uses.
No, it's a bug in third party code which couldn't be fixed by tdesktop.

>  I am an end-user of Telegram and not a developer

GitHub is a place where developers communicate, not end user support. You're supposed to file the bug to the right component on your own.

> Since your company is using Qt for your application, you should raise the issue with them. 

This is a community maintained bugtracker, not a company sponsored one. I'm not a Telegram employee.

> Perhaps Qt 6.7 relies on the presence of some other software on my system that it assumed were installed but is not? 

I don't know, I don't have expertise on fonts. The only thing I know is that you encountered a low level bug in Qt which tdesktop project has no resources to fix.

**Comment by hg2581** [comment]
> > Well, the bug is in Telegram regardless of which underlying toolkit it uses.
> 
> No, it's a bug in third party code which couldn't be fixed by tdesktop.
> 
> > I am an end-user of Telegram and not a developer
> 
> GitHub is a place where developers communicate, not end user support. You're supposed to file the bug to the right component on your own. You can contact end user support inside application
> 
> > Since your company is using Qt for your application, you should raise the issue with them.
> 
> This is a community maintained bugtracker, not a company sponsored one. I'm not a Telegram employee.
> 
> > Perhaps Qt 6.7 relies on the presence of some other software on my system that it assumed were installed but is not?
> 
> I don't know, I don't have expertise on fonts. The only thing I know is that you encountered a low level bug in Qt which tdesktop project has no resources to fix.

OK, understood. I ran CentOS 7 up until a little over a month ago and Chinese characters were rendered correctly in the Telegram Desktop. Since then two things happened: i upgraded from CentOS 7 to Rocky Linux 9, both RHEL compatible distributions of course and there was likely some upgrade on Telegram in between since I cannot remember which version I ran on the CentOS 7 system.

I also had a lot more software installed on my CentOS 7 system than I have on this comparatively fresh RL 9 system so it's quite possible there is some app/library missing on my RL 9 system that the Telegram Desktop implicitly relies on,

**Comment by ilya-fedin** [comment]
That's unlikely, all font-related libraries seem to be hard dependencies of Qt (i.e. the binary won't even run without them). More likely you had some other Chinese font installed, supported by Qt.

**Comment by hg2581** [comment]
> That's unlikely, all font-related libraries seem to be hard dependencies of Qt (i.e. the binary won't even run without them). More likely you had some other Chinese font installed, supported by Qt.

That is an interesting thought worth exploring - which CJK fonts are hard-coded by Qt? I could see if I could install one of them.

**Comment by ilya-fedin** [comment]
There's none. I guess there's something your fonts that Qt can't parse or something like that. I think noto fonts are the most widespread choice.

**Comment by hg2581** [comment]
> There's none. I guess there's something in your fonts that Qt can't parse or something like that. I think noto fonts are the most widespread choice.

Good, that was it. I found the following three noto-related libraries on epel:
google-noto-cjk-fonts-common
google-noto-sans-cjk-ttc-fonts
google-noto-serif-cjk-ttc-fonts

I began by just installing fonts-common and that was not enough. I then experimented and found that also installing noto-sans or noto-serif remedied the problem. Both Chinese text messages and the Korean language menu item now render correctly. Of course, I had to exit Telegram Desktop between each of the experiments.

I have to surmise that Qt hard-coded noto into the library since the default CJK font on RL 9 does not work. There ought to be a fall-back in Qt if the preferred font is not found - and all-squares should not be the fall-back.
